### DISCUSSION TITLE
Hosting a Secure Concourse Instance without VPN

### DISCUSSION THREAD

**Discussion by jeremyrcampb** [concourse/concourse#7010] [General]
I'm seeking a way to host a Concourse instance and securing its access without the need for a VPN.  I've been unsuccessful in finding a way to do that - and I'm looking for feedback/ideas on my approaches.

## Why no VPN?

A VPN is handy because you can permit access to your instance from a known list of trusted IP addresses.  However, we are trying to move away from this model and toward a zero-trust [BeyondCorp](https://beyondcorp.com/) model.

A key requirement for me is that I don't want to trust the application or the application runtime environment to protect from random baddies on the internet.  I want a layer in front of the application to prevent unauthorized access, exploit, DDoS, etc.  (More layers of security...)

### 1) Host behind HTTPS Proxy with Authentication (Google's Identity Aware Proxy / IAP)
This works for Browser based access, but fly doesn't work.  

Fly doesn't provide a Identity Token in the `Authorization: Bearer` on every call as Google's IAP requires, and when it does it is a identity token for Concourse, not for Google.

**_Maybe_** there would be a way to have something pre-populate the ~/.flyrc with a valid ID token, and maybe there would be a way to have the same ID token for both Google's IAP as well as for Concourse.  It would probably require Google's IAP and Concourse to use the same IDP, which they are not today.  In my case, Google's IAP is using Google Identity, which is federated with Okta.  And Concourse is using Okta, via OIDC.  This is theory, I don't know if it could pan out.

### 2) Two ATC instances, one behind IAP and one not.

In this design, I had one ATC instance behind the IAP at `https://url/`, and another behind a TCP load balancer at `https://fly.url/`, with the idea that the IAP endpoint would be fine for Browser-based access, which covers 90% of our use.  In the event `fly` is needed to be used, the alternative `fly.url` could be used, which may be protected either by an IP allow-list and only permitted from a bastion/known hosts.

This didn't work, it breaks the Concourse authentication flow.  Each ATC instance requires a `externalUrl` configured.  I tried two configs:

#### 2a) Both ATC instances configured with same `externalUrl`
When configuring both as `https://url/`, the fly auth workflow would get messed up due to browser redirects to that primary endpoint.

#### 2b) Both ATC instances configured with unique `externalUrl`
When both were configured individually as `https://url/` and `https:/fly.url/`, both worked, but not at the same time. Whichever's ATC instance came up most recently would work.  

The failing instance would error with: `Bad Request - Unregistered redirect_uri ("https://url/sky/callback").`

I assume this is because this is a single value configured in the database vs. a list of values, and the most-recently started pod would populate this with its configured value.

Protection via TLS client-certs, even if it did work via multiple ATC instances with unique URLs, is a non-option for us because we use the Terraform Concourse provider, which doesn't support client-certs (yet?).

### 3) Client TLS Certs

Requiring TLS client certs for connectivity to a single Concourse ATC endpoint seems workable, though the least user-friendly approach.  I have not yet attempted this.  We don't currently have the PKI to support this well.  Additionally, the Terraform Concourse provider doesn't support client TLS, so that's a blocker for us.  

## Summary

I've learned the deployment of the Concourse is pretty flexible, but as yet I've been unsuccessful in meeting my goal for a secure deployment.  

Criteria for that are:  1) non-dependent on IP allow-lists, 2) not dependent on the app/app runtime environment to protect against exploit.  

I'm super interested if any others have ideas for me on this.  Maybe I'm missing something simple.  Maybe I need to open a request to support multiple `redirect_url` configurations.

**Comment by funkypenguin** [comment]
We're using Concourse as a virtualservice behind an istio ingressgateway in our Kubernetes cluster. In our case, we _do_ use a VPN to access some internal-only resources (_we have an "internal" ingressgateway for VPN-only services, and an "external" ingressgateway for externall-exposed services_).

I've not tried it personally, but if you've got Istio in the mix, you can do [all sorts of clever authorizationpolicies](https://istio.io/latest/docs/reference/config/security/authorization-policy/) when permitting external access to your virtualservices, including, IIRC, a custom JWT/bearer auth from your OIDC provider.

Another option which you might investigate is deploying something like [oauth2_proxy](https://github.com/oauth2-proxy/oauth2-proxy) in front of ATC (_we use this to secure access to resources which don't provide their own authentication layer, behind a keycloak realm_)

Cheers!
D
